### ISSUE TITLE
Allow CRS WKT to represent the CRS without requiring reader to compare with grid mapping parameters

### ISSUE THREAD

**Issue by snowman2** [cf-convention/cf-conventions#222]
**Title:** Allow CRS WKT to represent the CRS without requiring reader to compare with grid mapping parameters
**Moderator:** ???  
**Moderator Status Review [last updated: YY/MM/DD]:** ???
**Requirement Summary:**

I propose the requirement be changed like so:

There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. *If both crs_wkt and grid mapping attributes exist, the attributes must be the same and grid mapping parameters should always be completed as fully as possible. As such, information from either one (or both) may be read in by the user without needing to check both. However, in those situations where the two values of a given property are different, the CRS information cannot be interpreted accurately and users should inform the provider so the issue can be addressed.* ~~, then the value specified by the single-property attribute shall take precedence.~~ For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element), *the value of this attribute cannot be interpreted accurately.*  ~~then the former, being the more specific attribute, takes precedence.~~ Naturally if the two values are equal then no ambiguity arises.

**Benefits:**
1. The CRS could originate from several different formats such as WKT, PROJ, or SRS Authority Code. If there are errors in the conversion process to the CF or WKT representation, only the provider would have the original CRS representation. As such, if there are conflicts, the provider would be the best source to go to in order to resolve the conflicts.
2. Making this change will simplify the lives of software developers so they can just read in the WKT or grid mapping CF parameters for the CRS without a need to compare the two.

**Status Quo:**
http://cfconventions.org/cf-conventions/cf-conventions.html#use-of-the-crs-well-known-text-format mentions

```    
There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence. Naturally if the two values are equal then no ambiguity arises.
```

**Comment by JonathanGregory** [comment]
The status quo (giving the CF attributes precedence over WKT) was discussed at great length when the possibility of including WKT strings was added. I have not reviewed that discussion but it would be relevant to do so to avoid repeating it! It's in https://cf-trac.llnl.gov/trac/ticket/69 and https://cf-pcmdi.llnl.gov/trac/ticket/80. I opposed the introduction of WKT strings because I didn't like redundancy, which would probably lead to inconsistency, but I agreed with the resolution that we have, in which the CF attributes take precedence.

Without reviewing the previous discussion, these points occur to me:

* The WKT model of metadata is different from the CF one. If I remember correctly, WKT overlaps with standard names and units of coordinate variables, not just with grid mapping attributes. Giving precedence to WKT would therefore mean that software would need to analyse the WKT in many circumstances. To enable this, we would need to draw up a thorough correspondence between the CF and WKT metadata models, with rules to resolve inconsistencies. That would be a big job

* We would have to revise these rules whenever the definition of WKT was amended - maybe this doesn't happen often, but it's not under our control, and we can't foresee what problems it might cause for CF.

* To give WKT precedence, we would have to require all CF-compliant applications to be able to parse WKT. That's a big expectation, which I think is unrealistic in practice.

In view of these points, I don't think this proposal is the best way to proceed. Instead, if there are elements of the CRS that can't currently be represented in CF but are needed, we should consider adding them, as we have done before (your points 1 and 3). If the equivalence between CF and WKT is unclear or incomplete (related to my first point above) it should be improved (your points 2 and 4).

**Comment by snowman2** [comment]
I am a GDAL/PROJ user, so from my biased perspective life would be much easier from the WKT form :). Additionally, since WKT is already a standard from the OGC geospatial community, most geospatial software should be able to support it.

> The WKT model of metadata is different from the CF one. ... That would be a big job.

Correct. That is why I propose the CRS WKT take precedence. The CF grid mapping parameters only provides support for a limited subset of projection parameters. (Ref: https://cf-trac.llnl.gov/trac/ticket/69):

> 3.2. Because the conceptual model for coordinate reference systems is both large and complex it is considered impractical to devise CF attributes for all of the potential CRS properties which might need to be encoded as metadata attributes in netCDF files. Consequently there is a requirement for such CRS properties to be specified in a compact notational format, preferably a format that is already in widespread use, either as a de facto or de jure standard.

So, in this proposal, if the CRS WKT exists and can be read in, the CF projection parameters should be ignored entirely and no checks made between the two. However, the CF projection parameters are there for both backwards compatibility as well as for programs that do not support the WKT form of the projection.

> To give WKT precedence, we would have to require all CF-compliant applications to be able to parse WKT. That's a big expectation, which I think is unrealistic in practice.

I should clarify that in this proposal that `CRS WKT` can remain *optional*. However, when it does exist and your program can read it in, I propose that it should take precedence. As a side note, with the GDAL Barn changes (https://gdalbarn.com/), reading in a WKT is much more practical with PROJ as a dependency. It also provides support for WKT2. Additionally, GDAL can easily support the WKT form of the projection which enables all the dependent software to read in the projection.

**Comment by snowman2** [comment]
Here is the WKT2 form of the `British National Grid` from: https://cf-pcmdi.llnl.gov/trac/ticket/80

```python
>>> from pyproj import CRS
>>> cc = CRS("OSGB 1936 / British National Grid")
>>> cc
<Projected CRS: EPSG:27700>
Name: OSGB 1936 / British National Grid
Axis Info [cartesian]:
- E[east]: Easting (metre)
- N[north]: Northing (metre)
Area of Use:
- name: UK - Britain and UKCS 49°46'N to 61°01'N, 7°33'W to 3°33'E
- bounds: (-9.2, 49.75, 2.88, 61.14)
Coordinate Operation:
- name: British National Grid
- method: Transverse Mercator
Datum: OSGB 1936
- Ellipsoid: Airy 1830
- Prime Meridian: Greenwich

>>> print(cc.to_wkt(pretty=True))
PROJCRS["OSGB 1936 / British National Grid",
    BASEGEOGCRS["OSGB 1936",
        DATUM["OSGB 1936",
            ELLIPSOID["Airy 1830",6377563.396,299.3249646,
                LENGTHUNIT["metre",1]]],
        PRIMEM["Greenwich",0,
            ANGLEUNIT["degree",0.0174532925199433]],
        ID["EPSG",4277]],
    CONVERSION["British National Grid",
        METHOD["Transverse Mercator",
            ID["EPSG",9807]],
        PARAMETER["Latitude of natural origin",49,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8801]],
        PARAMETER["Longitude of natural origin",-2,
            ANGLEUNIT["degree",0.0174532925199433],
            ID["EPSG",8802]],
        PARAMETER["Scale factor at natural origin",0.9996012717,
            SCALEUNIT["unity",1],
            ID["EPSG",8805]],
        PARAMETER["False easting",400000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8806]],
        PARAMETER["False northing",-100000,
            LENGTHUNIT["metre",1],
            ID["EPSG",8807]]],
    CS[Cartesian,2],
        AXIS["(E)",east,
            ORDER[1],
            LENGTHUNIT["metre",1]],
        AXIS["(N)",north,
            ORDER[2],
            LENGTHUNIT["metre",1]],
    USAGE[
        SCOPE["unknown"],
        AREA["UK - Britain and UKCS 49°46'N to 61°01'N, 7°33'W to 3°33'E"],
        BBOX[49.75,-9.2,61.14,2.88]],
    ID["EPSG",27700]]

```

The `coordinate system` and `area of use` currently don't have an equivalent in the CF conventions. The coordinate system is important to note as the axis order is taken into account in PROJ 6+ and GDAL 3+.

**Comment by dblodgett-usgs** [comment]
Dear @snowman2 --  
I agree with @JonathanGregory, that if things are missing from CF that are in WKT, they should be added. 

Maybe the core of your proposal is actually best made to the GDAL / PROJ project to modify default behavior when working with CF data? When different, a warning could be issued and the WKT used with preference?

Regards - Dave

**Comment by rmendels** [comment]
@snowman2  @dblodgett-usgs @JonathanGregory 
Second what Dave says.  The argument seems to be to break everything in the CF world so that GDAL will work better with netcdf files.  Why not improve GDAL  (supposedly based on a talk I heard the new,  I believe yet to be released version indeed does have better support).

GDAL is a great library and a lot of work has gone into it,  but its netcdf support has always been sketchy .  When I first was directed to it years ago,  it could only do 2-D files,  and would flip the data,  even  when the metadata clearly said the axes went in the other direction  (it just ignored the metadata attributes).  That problem lasted for a long time  (for all I know it still does this).  GDAL has had problems with greater than 3-D files,  forecast files, DSG files,  files that are part of the NCEI examples for sending in data,  some issues with time, and some of the newer features in netcdf4 files.  

Things that can improve CF are most welcome,  things that would potentially break most present CF based software should  have to make an awfully strong case for the benefits.

**Comment by snowman2** [comment]
Thanks all for the comments! My desire here is to unite the geospatial (OSGeo) and CF-conventions here to simplifying things when transitioning between the communities. 

Much of the inspiration for this thought came when attempting to match PROJ parameters to the CF conventions as documented [here](https://github.com/cf-convention/cf-conventions/wiki/Mapping-from-CF-Grid-Mapping-Attributes-to-CRS-WKT-Elements#table-3---common-projection-parameter-names). There are several parameters that do not match up and in several cases a grid mapping does not exist.This is problematic for users who wish to convert back and forth between the two. However, since PROJ supports reading in the WKT string, the full CRS can be properly represented in that manner and no information is lost. Additionally, the [PROJ FAQ](https://proj.org/faq.html#what-is-the-best-format-for-describing-coordinate-reference-systems) strongly discourages the use of PROJ strings to represent the CRS and instead recommends using the WKT string.

> Maybe the core of your proposal is actually best made to the GDAL / PROJ project to modify default behavior when working with CF data?

This would indeed be problematic and confusing for users of GDAL to change this as the behavior would differ from the CF spec. This is already done in `pyproj` and an issue exists due to users noting that the behavior differs from the spec [here](https://github.com/pyproj4/pyproj/issues/401).


### Alternative proposal?

Thoughts on stating in the spec that if the CRS cannot be properly represented using the CF grid mapping parameters, that the CRS WKT form is recommended as a fall back (noting of course that this may not be compatible with some software)? Also, it would be good to note for users to make an issue in this repo with their CRS WKT that cannot be represented using the grid mapping parameters so the CF spec can be updated accordingly.

**Comment by rmendels** [comment]
Would point out that the newest version of Proj4 in its latest incarnation just introduced a bunch of changes in how things are done,  and in the CRS.  See some of the discussion related to the `R` packages `sp` and `sf`.  It is breaking a lot of software.

**Comment by erget** [comment]
I agree with @dblodgett-usgs and @JonathanGregory - we already have a clear hierarchy that establishes which values have precedence over which ones in the case of conflict. Data producers already have the possibility of omitting CF attributes in favour of using WKT, although this is discouraged. I would see this as an acceptable solution if one wanted to produce data **now** and the relevant parameters weren't supported by CF. Optimally, one would pursue the adoption of the needed parameters in CF in parallel.

**Comment by rsignell-usgs** [comment]
This is a bit of a sidebar, but one thing that would make it easier for people EPSG and WKT folks to create the CF representation would be if we could get the friendly folks over at spatialreference.org to supply the CF representation.  If I google "EPSG 4326", I end up at https://spatialreference.org/ref/epsg/wgs-84/ 
![Opera Snapshot_2019-12-31_074229_spatialreference org](https://user-images.githubusercontent.com/1872600/71622034-38268900-2ba1-11ea-94c8-cd15ba12c90d.png)
which provides the WKT representation in several flavors as well as other representations.   Why not one more for CF?    This would also be a good way to figure out what is missing in CF...

**Comment by snowman2** [comment]
@rsignell-usgs, that would definitely be nice. However, it will also require a lot of work, so I imagine some kind of funding would be needed.

https://github.com/OSGeo/PROJ/issues/1193

**Comment by marqh** [comment]
Hello @snowman2 

this is an interesting topic and I am grateful that you have raised it

I think there are some fine details that are being picked out here that are interesting, as well as the big picture.

Whilst the big picture comes with a lot of considerations, there are small scale benefits we can try to get to.

One example stands out for me from your comments:

> The coordinate system is important to note as the axis order is taken into account in PROJ 6+ and GDAL 3+.

I have also been looking at the axis order with respect to CRS-WKT.  I agree that this is important.

I think that there is an in situ feature that can be extended to provide some extra clarity on this topic.

With this in mind, I have opened a new issue
#223 
 to discuss this as an isolated topic, to see if there is a quick and easy extension that would address this concern.

I very much support the broader scope discussion on this topic, hence my approach to separate out #223 so that the discussion on that targeted topic does not get in the way of these valuable considerations.

I hope this is a useful step
mark

**Comment by marqh** [comment]
> This is a bit of a sidebar, but one thing that would make it easier for people EPSG and WKT folks to create the CF representation would be if we could get the friendly folks over at spatialreference.org to supply the CF representation. If I google "EPSG 4326", I end up at https://spatialreference.org/ref/epsg/wgs-84/

Hi @rsignell-usgs 

I'm afraid that Google may be somewhat unhelpful with its advice

the resources at  https://spatialreference.org are not very well maintained, and the process of maintenance has been far from clear for some time:
https://spatialreference.org/about/

The EPSG maintain the official registry for EPSG codes, providing URI and URN notation for encodings, e.g.
https://www.epsg-registry.org/export.htm?wkt=urn:ogc:def:crs:EPSG::4326

Comparing this resource to 
https://spatialreference.org/ref/epsg/wgs-84/ogcwkt/
it is clear (to me) that the spatial reference resource have not adopted the updated CRS-WKT syntax, which is implemented in GDAL, ESRI and other modern packages.

At present the only well maintained resource for EPSG codes in WKT encoding that I am confident of using is
https://www.epsg-registry.org/

all the best
mark

**Comment by marqh** [comment]
On the detail point of the proposal, I would support amending the current text:

> There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence. Naturally if the two values are equal then no ambiguity arises.

To remove the latter precedence statement. 

> There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. ~~However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence. Naturally if the two values are equal then no ambiguity arises.~~

It is my view that there is too much of an onus placed on the data consumer here, to parse both content representations, map terms to one another and interpret outputs.  This is complicated and difficult to implement.  There are many opportunities for mistakes and problems.

If there is WKT in a file, I want my application to trust it, not to have to parse it to look for mistakes.  If i can just parse it then I can delegate this to a supporting application, which is great for maintainability.

I think that placing the onus on the data producer to produce content that they assert is consistent is sufficient.

I think the value of data consumers being able to simply parse the WKT directly is very large.

I think the cost of managing the assertion of consistency on data producers is much smaller.  In a sense the status quo is standardising for mistakes in encoding, which i don't think the standard should do, especially given the cost here.

all the best
mark

**Comment by snowman2** [comment]
This is also a useful resource: http://epsg.io/

**Comment by graybeal** [comment]
I read both [#69](https://cf-trac.llnl.gov/trac/ticket/69) and [#80](https://cf-pcmdi.llnl.gov/trac/ticket/80), and was startled by the sudden acceptance of these tickets after such long discussion of possible issues. (Credit here to Jonathan for flexibility!) Many of those issues are raised in this context, but this ticket proposes WKT be dominant in a much narrower sense (see detailed item (b) below).

I agree strongly with @margh's recent points, including the large value of data consumers being able to simply parse the WKT directly. It's key to recognize this is an augmentation, not a restriction. My detailed reasons follow, but first, I think the phrasing at the beginning of the proposal is creating unneeded alarm. 

Despite the misleading title, the proposal doesn't make WKT dominant, it just makes it directly usable (but still secondary, because the WKT is not required). I offer this as an equivalent rewrite of the proposal's first paragraph:

> I propose that if a CRS WKT is present and can be used by the software program, that the WKT should be allowed to stand alone as an official CRS of the file by CF standards (thus, implicitly ignoring non-WKT CRS parameters). However, non-WKT CRS parameters still must be present to serve as an official representation of the CRS, in the event the software program cannot read in the CRS WKT or chooses not to use it.

In the text you wouldn't say anything like this of course. The text already describes how WKT is an optional augmentation, and that the non-WKT CF must be as complete as possible. I'd only tweak one line, just before the paragraph marqh highlighted, by replacing "as well as by crs_wkt" with "even if a crs_wkt is present", so now it reads:

> Therefore the CRS should be described as thoroughly as possible with the single-property attributes, even if a crs_wkt is present.

With the [proposed precedence deletion of marqh](https://github.com/cf-convention/cf-conventions/issues/222#issuecomment-569938096) (item (c) below),  I believe this fully captures the intent of the proposal.

Detailed responses to a few points:

(a) Yes ideally CF could be equally capable. On the other hand, WKT will continue to improve and many tools are and will be built around it. Does CF want to take on the job of "keeping up with WKT" and expect tool developers to "keep up with the CF version of WKT"?  Even if we want that to happen, who in CF wants to volunteer to make it happen for CF? And in the tool community?
(b) This seems to be a graceful co-habitation strategy. I don't see how this "breaks everything in the CF world"—it is adding a capability to CF, not breaking anything that exists. If the data creator doesn't add WKT, it doesn't apply. If the tool reading the file doesn't support WKT, it doesn't apply. Everything presented in CF will continue to work with all the tools written for CF. If I'm trying to create CF-compliant data, even if my WKT adds critical value I will want to make my CF as complete and accurate as I can for non-WKT applications, and the CF parameters are still required by the convention.
(c) Supporting [the proposed precedence deletion of marqh](https://github.com/cf-convention/cf-conventions/issues/222#issuecomment-569938096): If the added WKT does not align with the CF, the data creator has introduced a bug. This can be corrected by social pressure (as is usually the case for any mistakes in the data) and does not require custom text in CF defining the "true meaning" (independent of the originator's intent).  The tool creator is motivated to make their tool maximally useful given available time, including whether to favor the WKT expression and whether to cross-check the two expressions. Co-existence works here also and does not damage CF (because the CF parameters still have to be there). 

I'm assuming positive assessments of the prevalence of WKT, its features, and its community support for upgrades. If you agree these are favorable indicators, then there are two ways to consider the options. (1) How good will this be for existing CF users going forward? Although maybe not many of them need WKT yet, it will be favorable on balance, with little or no downside that I can see.  And more broadly, (2) How much will this encourage/allow the geospatial community to easily adopt and use CF? I think it will be quite encouraging.

**Comment by snowman2** [comment]
@graybeal, thanks for clarifying! I used your clarified version as I think it does a much better job of capturing the intent of the proposal.

**Comment by JonathanGregory** [comment]
Although I'm watching this repository, and I contributed to this thread, GitHub has sent me only one of the contributions to this issue, namely the most recent (before this one, 10 h ago by @snowman2). Shouldn't I receive all of them by email? I depend on email to be informed that some discussion is taking place.

**Comment by JimBiardCics** [comment]
A few comments on the discussion to this point. I think the discussion is moving in the overall right direction. If seems to me that there was confusion at first between implementations and uses on the one hand and design and conventions on the other. I think we need to seriously consider how big a job it would be to "re-invent the wheel" by trying to add to CF, even piecemeal, all the parameters needed to represent all coordinate reference systems (CRSs). The vast majority of us are not geodesists. We need to acknowledge that this is a significant discipline that we know little about, and allow the experts in that field to be the experts. Let's use the standards they have developed rather than build an inferior substitute.

CF added the ability to specify a few projected coordinate systems. We clearly must continue to honor those for backward compatibility purposes, but let's not add any new ones. I think we should encourage the use of WKT CRS declarations going forward and focus on what might need to be added to CF to resolve ambiguities that might be present. If I understood correctly, @JonathanGregory thought there were possible issues. I didn't see any specifics given, but I'd rather try to clear those up than follow a "make our own" approach any longer.

I've worked with a few data providers that attempted to add grid_mapping variables to their netCDF files. The majority of them botched it. They would have been much better off if they could have copied and pasted a WKT string rather than try to figure out how to read CRS definitions and map elements to CF grid_mapping attributes.

**Comment by dblodgett-usgs** [comment]
Great strategy @JimBiardCics. Having contributed an attempted implementation to map CF conventions to WKT -- I know how error prone and hard it can be. Moving toward support of WKT as a fully fledged option within CF is unambiguously a good thing in my mind.

@marqh's suggested text changes make a ton of sense to me. 

Should we also add something that emphasizes the points about _"graceful co-habitation"_ ?

**Comment by snowman2** [comment]
> Should we also add something that emphasizes the points about "graceful co-habitation" ?

Are you thinking something along the lines of:

*"If both a CRS WKT and grid mapping parameters exist, it is assumed that they are equivalent. As such, either one may be used to represent the CRS of the file."*

**Comment by graybeal** [comment]
Or to deal with the edge cases and be consistent with our expectations: 

_"If both a CRS WKT and grid mapping parameters exist, it is assumed that they do not conflict. As such, information from either one (or both) may be used to represent the CRS of the file, recognizing that the grid mapping parameters should always be completed as fully as possible."_

**Comment by snowman2** [comment]
One minor addition:
*"If the CRS cannot be represented using the grid mapping parameters, using only the CRS WKT is allowed. However, some applications will not be able to read in the CRS WKT form."*

**Comment by JimBiardCics** [comment]
@snowman2 Are there any applications that actively read in and use the CF grid mapping parameters?

**Comment by snowman2** [comment]
> @snowman2 Are there any applications that actively read in and use the CF grid mapping parameters?

The only application I am aware of that does so is GDAL. However, it also checks for the WKT string and compares the two at present. I am not sure about other applications, but I assume there are based on the current cf-conventions. :man_shrugging:

**Comment by JimBiardCics** [comment]
GDAL is one more than I was aware of. I'm not aware of any others.

**Comment by rmendels** [comment]
I would suggest maybe going a little slow on this one,  because my guess is many people are not following this discussion,  so they will not be able to respond.  I have no idea what for example Panoply does,  the Coastwatch tools,  Thredds,  Seadas,  and some others I can think of,  and we should be very careful we don't needlessly break things.  That is why I objected to the original proposal where the WKT string would take precedence.  That would almost require that WKT strings be supported in software.

**Comment by pp-mo** [comment]
>  Are there any applications that actively read in and use the CF grid mapping parameters?

Not sure if I'm answering the right question here, but [Iris](https://github.com/SciTools/iris) ***definitely does*** explicitly interpret CF grid-mapping terms : we have explicit code for translating various types of grid-mapping to an [Iris 'coordinate_system']().  
Currently supporting types : 
```
albers_conical_equal_area
azimuthal_equidistant
lambert_azimuthal_equal_area
lambert_conformal_conic
lambert_cylindrical_equal_area
latitude_longitude
mercator
orthographic
polar_stereographic
rotated_latitude_longitude
stereographic
transverse_mercator
vertical_perspective
```
Most of these can also be produced as output.

For instance, we are currently working on making let Iris accept a geostationary grid with missing false_easting/false_northing parameters : https://github.com/SciTools/iris/pull/3628 

We currently don't support the WKT text, but could presumably consider this in future.

**Comment by dopplershift** [comment]
MetPy is reading these and converting them for our purposes of plotting in CartoPy. netCDF-java (cc @lesserwhirls) reads them and uses them to set up coordinate systems.

**Comment by snowman2** [comment]
Just to be clear, this proposal does not propose to remove or override the `grid_mapping` parameters. It's main goal is co-existence with the CRS WKT.

Proposed WKT string statement modifications (modifications in italics):
> The crs_wkt attribute is intended to act as a supplement to other single-property CF grid mapping attributes (as described in Appendix F); it is not intended to replace those attributes. If data producers omit the single-property grid mapping attributes in favour of the compound crs_wkt attribute, software which cannot interpret crs_wkt will be unable to use the grid_mapping information. Therefore the CRS should be described as thoroughly as possible with the single-property attributes as well as by crs_wkt. *If the CRS cannot be represented using the grid mapping parameters, using only the crs_wkt attribute is considered valid. *

>There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. *If both a crs_wkt and grid mapping attributes exist, it is assumed that they do not conflict. As such, information from either one (or both) may be used to represent the CRS of the file, recognizing that the grid mapping parameters should always be completed as fully as possible.* ~~However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence. Naturally if the two values are equal then no ambiguity arises.~~

**Comment by rmendels** [comment]
_If the CRS cannot be represented using the grid mapping parameters, using only the crs_wkt attribute is considered valid._

Sorry,  but to my untrained eyes this essentially means software that purports to be CF compliant will have to be able to deal with WKT.  It is very simple,  I have a transformation that can't be represented by present grid mapping parameters, I use the crs_wkt attribute,  based on the standard that says the file is CF-compliant  but a lot of software that has been the basis for CF-compliance will not be able to properly read that file.  I must be missing something here because I don't see anyway around this.  It will create CF-complaint files that present software will not be able to  properly read.  Which will mean either a lot of incompatible files or possibly a lot of work for the authors of the present software that are the backbone of CF-compliance.  

I am also concerned that not enough feedback has been coming from the authors of these software. Most telling was the inability of the people who have driven this to name such software (so that they are not aware of the possible implications of these changes),  and even more that this is being driven to benefit GDAL so that GDAL doesn't have to change  (read the original posts).  

I want to reiterate that it is not that I don't think going in this direction in the long-run is a good idea (that is a separate issue),  I just don't think the implications have been entirely thought through,  in particularly the ramifications for software developers who may not have the time and resources to deal with the changes.    Which is why I keep saying we should go slow on this , perhaps create a lot of sample files,  create ones that use CRS that presently can't be properly represented,  and see what kind of havoc it does or does not create.  Only after testing over as many programs as possible,  then see how this should be worded.

**Comment by davidhassell** [comment]
I broadly agree with @rmendels (https://github.com/cf-convention/cf-conventions/issues/222#issuecomment-572780372) and @JonathanGregory three points (https://github.com/cf-convention/cf-conventions/issues/222#issuecomment-569325165). This is not a bad idea, but it is also not one that has a trivial implementation, and so would need some work before it should be accepted. My interest is a general one, but also as developer of CF-compliant software: `cfdm` and `cf-python`. These libraries currently ignore WKT. I would be happy to so extend them so that they did, but that is probably not a 5-minute job, as there first need to exist the well-defined mapping between those parts of WKT that correspond to CF parameters.



Expanding on one of the aforementioned points:

- _We would have to revise (or consider revising) CF whenever the definition of WKT was amended_. On the face of it, this is the same issue that incorporating UGRID into CF has (https://github.com/cf-convention/cf-conventions/issues/153). However, the UGRID and CF communities overlap considerably, and it will be a **rule** that the UGRID experts consider CF when making changes, and also consider UGRID when CF changes. WKT will, I presume, evolve with no such reference CF.

Thanks,
David

**Comment by graybeal** [comment]
There are two perspectives in this thread (and the original WKT discussions) about what it means for optional 'containers' like WKT to be acknowledged as 'valid' in CF. In one view, the acknowledgement puts some or all of the responsibility for validating the container's content on CF, the CF validation software, and the CF tool developers. In the other view, it can be treated as a black box, with the responsibility for correctly including WKT content, verifying WKT content, and using WKT content relegated to the actual WKT providers and users.

I think the acceptance of WKT as an 'allowed' container in those original proposals implicitly adopted the second view. To my understanding no tests for WKT content validity have been added, and tools have not been required to change anything. The gist of the proposal seemed in line with that view, because WKTs would still be optional and the CF attributes would still be required to the maximum extent possible. Therefore, I think the best approach is to modify the proposal to make it fully consistent with the second view, because that lets CF dynamically and quickly adapt to advanced needs of an important community.

> _If the CRS cannot be represented using the grid mapping parameters, using only the crs_wkt attribute is considered valid._
> 
> Sorry, but to my untrained eyes this essentially means software that purports to be CF compliant will have to be able to deal with WKT. It is very simple, I have a transformation that can't be represented by present grid mapping parameters, I use the crs_wkt attribute, based on the standard that says the file is CF-compliant but a lot of software that has been the basis for CF-compliance will not be able to properly read that file. I must be missing something here because I don't see anyway around this. It will create CF-complaint files that present software will not be able to properly read. Which will mean either a lot of incompatible files or possibly a lot of work for the authors of the present software that are the backbone of CF-compliance.

I agree that the quoted sentence is ambiguous and inconsistent with other text. I'd prefer it went away; or at least be expressed as "If the CRS cannot be fully represented using the grid mapping parameters, the additional crs_wkt attribute can be used to augment the representation (while noting that support for reading and using the crs_wkt attribute in CF software remains entirely optional)."

Still, there are oodles of CF-compliant data files that present software can't "properly" read, because they have various optional augmentations in them not understood by all software. The fact this particular optional augmentation is mentioned explicitly does not mean everyone, or even anyone, has to support it. It also does not mean that any testing or conformance software *has to* support or test WKT, any more than allowing multiple conventions in the convention attribute means you have to support and test all of the conventions that might be referenced by that attribute.

> We would have to revise (or consider revising) CF whenever the definition of WKT was amended. 

I'm not sure why. The only thing the WKT content can affect is another CF tool that reads WKT. All the existing tools that ignore WKT should not be impacted, they'll ignore WKT no matter what its definition is. Now, the tools that use WKT in CF might have to decide how to deal with WKT versions, but they could do that without ever having to require changes to the CF specification.

**Comment by steingod** [comment]
> I think the acceptance of WKT as an 'allowed' container in those original proposals implicitly adopted the second view. To my understanding no tests for WKT content validity have been added, and tools have not been required to change anything. The gist of the proposal seemed in line with that view, because WKTs would still be optional and the CF attributes would still be required to the maximum extent possible. Therefore, I think the best approach is to modify the proposal to make it fully consistent with the second view, because that lets CF dynamically and quickly adapt to advanced needs of an important community.

I think this is a beneficial approach.

**Comment by JimBiardCics** [comment]
@graybeal I am also in favor of the "black box" approach to WKT. We should accept the work of the experts that develop WKT and define CRSs using it, and trust the registered WKT strings for different CRSs. Someone can always produce an incorrect WKT CRS string, but someone can also produce an incorrect grid mapping variable. The CF checker doesn't try to verify the exact contents of grid mapping variables, so I don't see why we should involve ourselves with verifying the exact contents of WKT CRS strings.

(Mild rant alert)

The great majority of CF files don't contain grid mapping information. A number of  those that do contain incorrect information — either improperly chosen attributes or incorrect values for the attributes. (Based on the small sample of datasets that I've QC'd, most of the producers who add grid mapping get it wrong.) Tools that are able to use grid mapping attributes and attempt to use any of these files will not produce "best" results, and will in some cases produce crazy results. I am therefore not overly swayed by the appeal to issues for existing software. Existing tools that use grid mapping variables do so because that was the mechanism which was provided. If we had provided WKT or Proj strings from the beginning, those are what existing software would be using.

I confess that I haven't done the work to prove it, but I believe that the CF grid mapping attributes are insufficient to fully represent most CRSs, even the ones that we claim to support. They are sufficient if your spatial accuracy and precision is rougher than a kilometer or so, but if you are are concerned about meter-level (or better) accuracy in all three spatial dimensions, I don't think we support enough of the attributes needed to do so.

I think we should be encouraging software tool developers to embrace WKT (and/or Proj) CRS declarations instead of viewing them as second-class citizens in CF-land. I'd love to see a day arrive when we decided to deprecate grid mapping attributes for new files.

**Comment by Dave-Allured** [comment]
Moving toward WKT would be facilitated by a few software modules that interpret WKT and translate to familiar CF grid_mapping parameters, when reasonably compatible.

@snowman2 wrote:
> with the GDAL Barn changes (https://gdalbarn.com/), reading in a WKT is much more practical with PROJ as a dependency. It also provides support for WKT2. Additionally, GDAL can easily support the WKT form of the projection which enables all the dependent software to read in the projection.

"enables all the dependent software to read" -- Which language API's are supported or planned for this?  Are there known efforts or interest to develop some WKT-to-grid_mapping translation modules?

**Comment by snowman2** [comment]
PROJ is in c/c++. Currently available bindings for other languages: https://proj.org/development/bindings.html

**Comment by snowman2** [comment]
> Are there known efforts or interest to develop some WKT-to-grid_mapping translation modules?

I brought this up in the past: https://github.com/OSGeo/PROJ/issues/1193. However, I think it would require more interest in the community and/or funding to make it happen.

I have a basic implementation based on the [PROJ strings mappings](https://github.com/cf-convention/cf-conventions/wiki/Mapping-from-CF-Grid-Mapping-Attributes-to-CRS-WKT-Elements) in `pyproj`:
- https://pyproj4.github.io/pyproj/stable/api/crs.html#pyproj.crs.CRS.from_cf
- https://pyproj4.github.io/pyproj/stable/api/crs.html#pyproj.crs.CRS.to_cf

However, the mapping back and forth is imperfect and misses quite a bit. I have plans to update grid mappings based on WKT as that will allow a more complete mapping. PROJ string mappings code [here](https://github.com/pyproj4/pyproj/blob/67ff137f8d734fbaa4aac32363b85dd5e33dad5c/pyproj/cf1x8.py).

**Comment by JonathanGregory** [comment]
I'm not contributing much to this because GitHub is inexplicably not sending me the contributions to this issue. It may have decided for itself that it's better for all if I'm kept in the dark!

I feel that the best solution for this would be for CF to maintain a document which describes the mapping between CF metadata and WKT, and for conformance to this mapping to be checked by the CF-checker. In that way a user could be reasonably confident of getting the same result by reading either of them, and errors would be identified. I think we already have some documents somewhere that are relevant (prepared years ago by Etienne Tourigny).

This would be a non-trivial piece of work but until it is done I don't think we can really know what we are missing in the CF attributes what WKT can describe and needs to be included in use cases of CF datasets. It could be that CF `grid_mapping` attributes could eventually be replaced by WKT, as Jim would prefer, but I suspect that there is overlap with other parts of the CF data model, which would make this hard. It's certainly worth considering, but I think that writing down the mapping is a necessary piece of evidence in deciding about that and would therefore be a useful first step.

**Comment by snowman2** [comment]
@JonathanGregory whenever you see this :smile:, I would recommend checking the right hand side of this issue at the top to see if your subscribed to notifications:
![image](https://user-images.githubusercontent.com/8699967/72623135-2e738400-390a-11ea-99ab-345ce13d5d4f.png)

**Comment by snowman2** [comment]
> I think that writing down the mapping is a necessary piece of evidence in deciding about that and would therefore be a useful first step.

Here are some examples of the WKT version for the CF grid_mappings for the conversions:
https://pyproj4.github.io/pyproj/latest/_modules/pyproj/crs/coordinate_operation.html

This may also be useful:
https://pyproj4.github.io/pyproj/latest/build_crs.html

**Comment by rschmunk** [comment]
@rmendels said

> I have no idea what for example Panoply does, the Coastwatch tools, Thredds, Seadas, and some others I can think of, and we should be very careful we don't needlessly break things.

Panoply uses the netCDF-Java library to read datasets. When dealing with projected grids, it looks for the variable named by the `grid_mapping` attribute and the CF specified projection parameters of that variable. For the 13 projections that Panoply supports, in 4 cases it uses projection code that is part of NJ but in the other 9 it uses its projection code to transform the grid. The reason for those 9 is that I'd already implemented them for other purposes, and the NJ library doesn't do the best job of performing sanity tests on the projection attributes that it finds in the dataset.

FWIW, I only started looking through this thread because an issue regarding WKT was opened at Unidata/netcdf-java#191 earlier today.

**Comment by snowman2** [comment]
> This would be a non-trivial piece of work but until it is done I don't think we can really know what we are missing in the CF attributes what WKT can describe and needs to be included in use cases of CF datasets. 

@JonathanGregory, this is currently supported in `pyproj` master: https://github.com/pyproj4/pyproj/blob/6390041fc2bfb569e90f615adb7ba75160d41aa8/pyproj/crs/_cf1x8.py

You can see the tests that were run to check the conversion to/from the CF format of the projection [here](https://github.com/pyproj4/pyproj/blob/6390041fc2bfb569e90f615adb7ba75160d41aa8/test/crs/test_crs_cf.py).

I didn't find any parameters that weren't able to be supported by WKT. However, there are many projections that WKT can describe that aren't supported by CF. There is an extensive list [here](https://proj.org/operations/projections/index.html) for projections supported by PROJ. Only a subset is supported by CF conventions.

**Comment by snowman2** [comment]
I am removing the part about the standalone WKT to be discussed in another issue. 

Proposed WKT string statement modifications (modifications in italics):
>There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. *If both a crs_wkt and grid mapping attributes exist, it is assumed that they do not conflict. As such, information from either one (or both) may be used to represent the CRS of the file, recognizing that the grid mapping parameters should always be completed as fully as possible.* ~~However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence. Naturally if the two values are equal then no ambiguity arises.~~

**Comment by graybeal** [comment]
First two sentences are duplicated in the second two sentences?

**Comment by snowman2** [comment]
Good catch! Just fixed.

**Comment by JonathanGregory** [comment]
I continue to think that if we want to allow `grid_mapping` attributes and WKT to have equivalent status, we should write down the mapping between them, so it can be checked automatically. Thanks for pointing out the work on this in `pyproj` - is that your work? Without being able to do this check, I fear that some data-writers who are solely interested in one representation and not familiar with the other will very likely make mistakes and the file will be incorrect. I think this is too unsafe for us to rely on a stated obligation to get it right.

If you have use cases for actual datasets that you want to put in CF but the required projections are not supported in `grid_mapping`, it would be useful to propose adding them to CF.

By the way, my problem with not receiving all CF GitHub posting turned out to be that gmail thought some of them were spam, but it wasn't consistent about that - it objected to some comments and not others.

**Comment by rmendels** [comment]
+1 On all points

As I have said before, it is not that this isn't a good idea,  but why not take the time to do it correctly with minimal impacts.

**Comment by rsignell-usgs** [comment]
@rschmunk and @lesserwhirls, could panoply and netcdf-java use the Java [geotools](https://github.com/geotools/geotools) library to handle [CRS WKT](https://docs.geotools.org/stable/userguide/library/referencing/crs.html)?

Would that be straightforward or a major effort?

**Comment by lesserwhirls** [comment]
> @rschmunk and @lesserwhirls, could panoply and netcdf-java use the Java [geotools](https://github.com/geotools/geotools) library to handle [CRS WKT](https://docs.geotools.org/stable/userguide/library/referencing/crs.html)?
> 
> Would that be straightforward or a major effort?

It would introduce a somewhat interesting circular dependency (they are using netCDF-Java 4.6.11, we're on 5.2.0). I've been looking at `proj4j` to see what they have for parsing WKT, at which point we'd need to setup a mapping to CF parameters and create a netCDF-Java Projection object. We have some (what looks to be) contributed code that does the parsing, mapping, and projection object creation, but it's not very forgiving in its parsing and not super capable at this point.

**Comment by JimBiardCics** [comment]
@JonathanGregory said
> If you have use cases for actual datasets that you want to put in CF but the required projections are not supported in grid_mapping, it would be useful to propose adding them to CF.

I personally have a number of cases where CF doesn't provide the projection used. I come back to the question of why we feel the need to re-invent the wheel. I believe we need to learn to coexist with other standards and conventions that are at least as stable and reliable as CF rather than replicate them.

**Comment by JonathanGregory** [comment]
Dear @JimBiardCics

I appreciate the point and I agree that we should not spend our time doing a less good job of something which is already available elsewhere. However I don't think that CF is fairly described as "reinventing the wheel" in this case. I can suggest a few reasons why we have our own set of `grid_mapping` definitions rather than relying on `proj4` for example.

* An external resource might not include all the cases we need. Does `proj4` include rotated pole and vertical perspective, for instance? (If it does that is not noted in our App F). If it's not our own resource, we can't necessarily add to it, and if it's not comprehensive, users of CF might have to mix different conventions in their files, which would be less satisfactory.

* An external resource might change in a way which didn't suit CF, meaning we'd have to do extra work in a hurry to replace it with something else, with the further penalty of backwards-incompatibility for CF.

* An external resource might not present the information in a way which is easily usable in CF, because it's not organised in a similar fashion or doesn't match well to the CF data model.

Therefore instead of "handing over" something to an external authority, I feel that in most cases it is better to import the external information to CF. This isn't an absolute rule, of course, but it seems fine for `grid_mapping` to me. However, your point is a good reason why we should make the choice of attributes as similar as we can to `proj4`. If we do that, it should be easy to import more of them, like new standard names on existing patterns, and it improves interoperability between conventions. If we find there is a use case which is *not* trivial to import, there's probably a good reason for that which will require some thought, like new standard names which propose new patterns.

Best wishes

Jonathan

**Comment by snowman2** [comment]
> Does proj4 include rotated pole and vertical perspective, for instance? 

Just to have it noted that it is supported in PROJ (the conversion classes will be released in pyproj 2.5.0 soon, but the PROJ string version works on all versions):
```python
>>> from pyproj.crs.coordinate_operation import RotatedLatitudeLongitudeConversion, VerticalPerspectiveConversion
>>> RotatedLatitudeLongitudeConversion(o_lat_p=1, o_lon_p=2).to_proj4()
'+proj=ob_tran +o_proj=longlat +o_lat_p=1 +o_lon_p=2 +lon_0=0'
>>> VerticalPerspectiveConversion(viewpoint_height=50000).to_proj4()
'+proj=nsper +lat_0=0 +lon_0=0 +h=50000 +x_0=0 +y_0=0'
```

From tests:
```python
    crs = CRS.from_cf(
        dict(
            grid_mapping_name="rotated_latitude_longitude",
            grid_north_pole_latitude=32.5,
            grid_north_pole_longitude=170.0,
        )
    )
    expected_cf = {
        "semi_major_axis": 6378137.0,
        "semi_minor_axis": crs.ellipsoid.semi_minor_metre,
        "inverse_flattening": crs.ellipsoid.inverse_flattening,
        "reference_ellipsoid_name": "WGS 84",
        "longitude_of_prime_meridian": 0.0,
        "prime_meridian_name": "Greenwich",
        "horizontal_datum_name": "World Geodetic System 1984",
        "grid_mapping_name": "rotated_latitude_longitude",
        "grid_north_pole_latitude": 32.5,
        "grid_north_pole_longitude": 170.0,
        "north_pole_grid_longitude": 0.0,
    }
    cf_dict = crs.to_cf()
    assert cf_dict.pop("crs_wkt").startswith("PROJCRS[")
    assert cf_dict == expected_cf
```
```python
    crs = ProjectedCRS(conversion=VerticalPerspectiveConversion(50, 0, 1, 0, 2, 3))
    expected_cf = {
        "semi_major_axis": 6378137.0,
        "semi_minor_axis": crs.ellipsoid.semi_minor_metre,
        "inverse_flattening": crs.ellipsoid.inverse_flattening,
        "reference_ellipsoid_name": "WGS 84",
        "longitude_of_prime_meridian": 0.0,
        "prime_meridian_name": "Greenwich",
        "horizontal_datum_name": "World Geodetic System 1984",
        "grid_mapping_name": "vertical_perspective",
        "perspective_point_height": 50.0,
        "latitude_of_projection_origin": 0.0,
        "longitude_of_projection_origin": 1.0,
        "false_easting": 2.0,
        "false_northing": 3.0,
    }
    cf_dict = crs.to_cf()
    assert cf_dict.pop("crs_wkt").startswith("PROJCRS[")
    assert cf_dict == expected_cf
```

> If it's not our own resource, we can't necessarily add to it

It is an open-source tool that accepts contributions. Are you not allowed to contribute to other open-source projects where you work?

**Comment by taylor13** [comment]
I haven't been following this, but what about tripolar grids, which are often used in ocean models?

**Comment by zklaus** [comment]
How are they represented in CF right now? As far as I know only by 2d coordinates (which doesn't codify the iso-coordinate lines).

**Comment by JonathanGregory** [comment]
Dear @snowman2 
That's good about `proj` supporting those two projections. Thanks.
I used `proj` as an example but I was really thinking more about the general point that Jim raised, about relying on external resources. We've often discussed this on the former CF email list. I understand from what you say that `proj` is software, rather than a conventions document, so that's a bit different. You're right of course that open-source software can be modified by anyone for their own use. However, if someone proposes a new `grid_mapping` in CF, they just want the document to be modified, and they would not be offering to modify `proj` software to support this new mapping.
Best wishes
Jonathan

**Comment by JimBiardCics** [comment]
@JonathanGregory Am I right in understanding that back in the day the whole grid_mapping scheme in CF was added to annotate the relationship between lat and lon variables (which were always required) and X and Y variables? If the purpose was annotation and that annotation was itself optional, there was no actual need for the grid_mapping attributes provide sufficient information to do the transformation.

**Comment by JonathanGregory** [comment]
@JimBiardCics The `grid_mapping` attributes were introduced to allow software to do the transformation if it wanted to, for instance to transform other locations (apart from the coordinates and boundaries) or do other computations (such as working out local directions or cell areas).  They are optional because generic software does not need to be able to do the transformation if the 2D lat and lon coordinates are required to be present.

**Comment by pp-mo** [comment]
> @taylor13  I haven't been following this, but what about tripolar grids, which are often used in ocean models?
> @zklaus How are they represented in CF right now? As far as I know only by 2d coordinates (which doesn't codify the iso-coordinate lines).

From our specific experience : Most ocean models use the so-called ORCA grids.
These are not "just" a tripolar coordinate grid but use irregular sampling in various places.

Most importantly : ***a grid is not a projection***

So, a grid like ORCA has 2D latitude and longitude values for cell corners (and centres).  So that might *appear* to be like a coordinate-like space described by the grid indices.  But this is misleading, because a projection would be a smooth, reversible, total function  :  Instead ...
  1. you cannot define a point on the globe for an "interpolated" point in index-space like "ix=100.5, iy=101.3".  The relationship is entirely defined by the (2d) X and Y values, there is no smooth function from (ix, iy) to (lat, lon).
  1. there is no reverse mapping : you can't ask "where in the grid" a given point on the globe is.  Only which cells it may fall within... 
  1. in fact not all parts of the globe are covered at all.  The bottom of the space can be distorted to give extra detail in the Southern ocean, and does not cover the South pole.   Also, some model cells actually *overlap* along the top seam, referred to as a 'north fold'.  Thus...
  1. ... returning to the 'reversal' question, a given global location can be in several gridcells, or none.

In practice ORCA grids are particularly extreme case, in that the standard grids are defined _entirely_ by the provided 2D arrays of latitudes + longitudes, and there is no provided equation defining those numbers, as far as I can determine.

**Comment by davidhassell** [comment]
Hello,

The timings and order of the breakout groups for the CF meeting next week has now been set (see http://cfconventions.org/Meetings/2020-Workshop.html), and the discussion of this issue will be on Thursday 11 June from 16:00-17:30 UTC, in parallel with three other topics.

Thanks.

**Comment by JonathanGregory** [comment]
We're going to discuss this issue today. I've just written the following in the google doc about it, because I can't say all this in the discussion, and I would like to explain where I'm coming from when we talk.

The current position is that the `grid_mapping` takes precedence over WKT if both are present. To know whether this the case, the data-user would have to interpret both and compare them, and take the `grid_mapping` as correct if the two are inconsistent. If I understand correctly, the idea of the proposal is to remove the requirement of precedence, so that the data-user can read the WKT alone and not consider the `grid_mapping`. In fact, I think the data-user can behave in just that way as things stand. The data-user is not obliged to check that the dataset is self-consistent. They are entitled to assume that it is, because the onus is on the data-producer to ensure this. I believe that's what we had in mind when we decided the current convention, which was a sort of compromise. The statement of precedence is intended to resolve problems if they are detected. Hence I don't think we need to change the convention to achieve the aim stated (Allow CRS WKT to represent the CRS without requiring comparison with grid mapping Parameters), if I've understood it correctly. It's already OK.

The discussion in this issue has gone further than that, however. A lot of it is about whether the `grid_mapping` should be able to represent the information in the WKT. I think this is a much more difficult question. As I've commented in the issue, I have serious concerns about this. I realise that other people have good reasons for suggesting it, and I hope we can find a consensus, as usual - honestly, I do! I have two kinds of concern, which are related.

  * If `grid_mapping` isn't required to represent everything, data-writers may chose to omit it. Some people would like that, because it would permit them to write WKT in the CF-netCDF file and allow it to be read and used as a “black box” required by certain software. I can see the practical advantage in that, but I think that's contrary to the general intention of CF, which has made it successful and popular. CF metadata is intended to be self-explanatory and intelligible to humans, designed to be read and written with minimal possibilities for mistakes. WKT looks to me more like code, it's not so self-explanatory because it doesn't label all its parts, the order in which things appear determines what they mean. It's not like CF-like.

  * I suspect that WKT may be inconsistent with the CF data model in some ways. I can't be sure unless we understand it thoroughly and how it relates to CF metadata. For instance, @snowman2 noted in the issue that, “The coordinate system and area of use currently don't have an equivalent in the CF conventions.” I don't know what they are exactly, but I think that coordinate system is an idea which is inconsistent with the CF data model, as an explicitly recognised part of the metadata. I recall also (but perhaps incorrectly) that there are defaults about directions and units, which may be inconsistent with CF.

Because of these concerns, I continue to think that we should require data-procedures to describe everything they want to in `grid_mapping`, unless we can write down explicitly the mappings between CF metadata and WKT. We don't have to be able to carry out the conversion in software, but we should write down how to do it. When that has been done, we would have much greater clarity. We could be confident in stating that certain part of WKT were equivalent to CF metadata. It's possible that there are vocabularies in WKT which could be adopted by CF, and thus not maintained by CF. That idea is often suggested (e.g. by Jim in this issue) and it makes sense provided the vocabulary is one which can exactly “slot” into CF metadata. That is, it must suit our data model.

It seems to me that this approach of mapping is what is being followed with the discussions about standard names and ontologies. It is not being suggested that standard names should be replaced by vocabularies from elsewhere, for the same kind of reason that there isn't a one-to-one correspondence. However, the mappings can be established to help with interoperability.

But having written down the mapping between CF metadata and WKT, we would be able to write software that can do the conversion completely (@snowman2 already has such software, and probably others do). That could be used in the cf-checker to check that the `grid_mapping` and WKT are consistent, which is the problem we started with in this issue.

**Comment by snowman2** [comment]
> The statement of precedence is intended to resolve problems if they are detected. Hence I don't think we need to change the convention to achieve the aim stated (Allow CRS WKT to represent the CRS without requiring comparison with grid mapping Parameters), if I've understood it correctly. It's already OK.

@JonathanGregory - does this mean you approve of the wording change?

If everyone approves of the wording change, then we could re-purpose the meeting to another CRS WKT topic, such as some issues you have mentioned in your post.

**Comment by JonathanGregory** [comment]
If I've understood your concern correctly, I don't see the problem with the existing wording. It's intended to give guidance for what to do if a data-user notices an inconsistency. Is the problem that it can be understood to imply that the data-reader must read both versions and compare them? If so, maybe we could address that concern by keeping the present wording, but adding some more words to say that the data-user can assume they are consistent, and therefore needs to read only one of them. Jonathan

**Comment by snowman2** [comment]
> Is the problem that it can be understood to imply that the data-reader must read both versions and compare them?

Yes, there are several data readers that have understood it this way. So, clarification on this would be a welcome addition.

**Comment by snowman2** [comment]
Sounds like we should target the discussion towards the wording of the clarification to ensure all parties are on-board.

**Comment by snowman2** [comment]
> But, if the user wants to check the values for consistency, then when they conflict, the user should assume the CF value is correct. This gives the data writer a little extra incentive to try to get things right when converting the WKT metadata to CF.

I think that if the values are not consistent, it does not make sense to default to CF or WKT as either one or both could be incorrect depending on your starting format of the CRS and how the conversion was done.

In the proposal, it adds this clarification:
> *If conflicts exist between the representations, you should inform the provider so they can be addressed.*

Reasoning for this wording mentioned in the proposal:

> The CRS could originate from several different formats such as WKT, PROJ, or SRS Authority Code. If there are errors in the conversion process to the CF or WKT representation, only the provider would have the original CRS representation. As such, if there are conflicts, the provider would be the best source to go to in order to resolve the conflicts.

**Comment by taylor13** [comment]
I agree that informing the data provider of conflicts when found is good practice.  But what behavior should software have?  Should it just stop, or can we tell it which of the two conflicting pieces of information it should rely on (until the conflict has been resolved)?

My general ignorance about WKT prevents me from understanding what is meant by "then the value specified by the single-property attribute shall take precedence."   Is the single-property attribute sometimes the WKT property and sometimes the CF attribute, or is it invariably the CF attribute?

**Comment by JonathanGregory** [comment]
I agree that the data-producer should be the best authority on what was intended. However, knowing that doesn't give the data-user an immediate solution to an inconsistency. I think the current wording (the precedence of CF metadata over WKT) makes sense, since this is a CF dataset. As Karl says, that default also gives an incentive to the data-producer to ensure consistency. However, I think it's also fine to recommend contacting the data-producer.

**Comment by snowman2** [comment]
The final wording from the breakout meeting:

There will be occasions when a given CRS property value is duplicated in both a single-property grid mapping attribute and the crs_wkt attribute. In such cases the onus is on data producers to ensure that the property values are consistent. *If both crs_wkt and grid mapping attributes exist, the attributes must be the same and grid mapping parameters should always be completed as fully as possible. As such, information from either one (or both) may be read in by the user without needing to check both. However, in those situations where the two values of a given property are different, the CRS information cannot be interpreted accurately and users should inform the provider so the issue can be addressed.* ~~, then the value specified by the single-property attribute shall take precedence.~~ For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element), *the value of this attribute cannot be interpreted accurately.*  ~~then the former, being the more specific attribute, takes precedence.~~ Naturally if the two values are equal then no ambiguity arises.

**Comment by taylor13** [comment]
Which parts of the sentence:

"If both crs_wkt and grid mapping attributes exist, the attributes must be the same and grid mapping parameters should always be completed as fully as possible."

should trigger an error (warning?) in a compliance checker?  Is the file compliant if the two are *not* the same?  Is a file compliant if the grid mapping parameters are incomplete (when it is possible for them to be complete)?

**Comment by snowman2** [comment]
> should trigger an error (warning?) in a compliance checker?

I would say an error.

> Is the file compliant if the two are not the same?

I would say no based on this part: "in those situations where the two values of a given property are different, the CRS information cannot be interpreted accurately"

> Is a file compliant if the grid mapping parameters are incomplete (when it is possible for them to be complete)?

I would say no based on this part: "the attributes must be the same and grid mapping parameters should always be completed as fully as possible"

**Comment by cameronsmith1** [comment]
How about an entirely different solution:  When there are multiple grid descriptions in a file, the creator must add a metadata flag that indicates which of the grid descriptions is 'primary'. 

The onus to make grid descriptions as equivalent as possible can still be on the creator, but the user will know which one to trust if there is a discrepancy.  And the CF checkers will only need to check that there is one, and only one, 'primary' flag.

**Comment by snowman2** [comment]
That would require all applications to be able to support the WKT format. Currently that is not possible due to software limitations (see comments in this thread).

**Comment by JonathanGregory** [comment]
Dear all

Before the meeting yesterday I was arguing, like Karl @taylor13, to retain the presumption that if the `grid_mapping` and `crs_wkt` are inconsistent, the data-user should presume the `grid_mapping` is correct. I was persuaded by the discussion that this isn't generally helpful, because the data-user might well think it was unsafe to proceed on that basis. In particular, the data-user might be aware that the WKT information had come first, and therefore suspect the `grid_mapping` of being an incorrect translation. Karl and I had argued also that the presumption of `grid_mapping` being correct gives an extra incentive to the data-producer to make sure the two are consistent. However, this also probably doesn't work; if the data-producer had thought about the consequence of misinterpretation, they would have tried to avoid inconsistency anyway.

Therefore I support the change to remove this assumption, and state that the metadata is invalid if `grid_mapping` and `crs_wkt` are inconsistent. In response to Karl, I agree with Alan @snowman2 that this is an error, and the file is not compliant, because the convention states the two kinds of metadata must be consistent.

Unfortunately, the CF checker won't be able to detect this error unless we write down the mapping between `grid_mapping` attributes and `crs_wkt` in the conformance document (or some document it can refer to). To make the check, you have to be able to interpret both kinds and compare them. One of Alan's concerns was that data-users felt they had to do that. We agreed that they don't. They can read one or the other, assuming they agree. However, it would be good if this could be checked routinely. As I argued before, I strongly feel that it would improve the convention if we could write down the equivalence. Note that we don't have to consider _all_ aspects of WKT, but only those aspects which people want to write in CF-netCDF files. Doing this would cause us to identify what has to be added to `grid_mapping` to give it the required capabilities, and whether there are inconsistencies between the CF data model and WKT. I wouldn't be surprised if there are, and we need to know, because it's not safe to treat `crs_wkt` as a black box if it might conflict with other CF metadata (not just in the `grid_mapping`, but for instance in the units and standard name of coordinate variables).

I would be concerned about adopting Philip @cameronsmith1's suggestion, because I fear that might lead to data-producers not being so careful with one or the other of the representations, thinking that they could set the flag to indicate it's not to be trusted.

Jonathan

**Comment by JonathanGregory** [comment]
This issue doesn't have a moderator - I think that's why it's not progressed. I will moderate it. @snowman2's current proposal (https://github.com/cf-convention/cf-conventions/pull/282) is to replace

However, in those situations where two values of a given property are different, then the value specified by the single-property attribute shall take precedence. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element) then the former, being the more specific attribute, takes precedence.

with

If both crs_wkt and grid mapping attributes exist, the attributes must be the same and grid mapping parameters should always be completed as fully as possible. As such, information from either one (or both) may be read in by the user without needing to check both. However, in those situations where the two values of a given property are different, the CRS information cannot be interpreted accurately and users should inform the provider so the issue can be addressed. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis and also by the crs_wkt attribute (via the WKT SPHEROID[…​] element), the value of this attribute cannot be interpreted accurately.

I think this is OK, except for the last sentence, which I think should be

For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis **disagrees with** the crs_wkt attribute (via the WKT SPHEROID[…​] element), the value of this attribute cannot be interpreted accurately.

That leads naturally to the unaltered final sentence, "Naturally if the two values are equal then no ambiguity arises."

Philip @cameronsmith1 and Karl @taylor13, are you content with this? Alan @snowman2, is my amendment OK with you?

Jonathan

**Comment by taylor13** [comment]
Yes, I think the intent of this is fine.  

I don't recall the text that precedes the revised text, but should the first sentence read:

"If, *for a given property,* both crs_wkt and grid mapping attributes exist, the attributes must be the same and grid mapping parameters should always be completed as fully as possible"

Also, is the second clause dependent on the first clause, or in general is it true that "grid mapping parameters should always be completed as fully as possible."  If it is generally true, I think the second clause should be its own sentence (and maybe it should appear elsewhere?).

**Comment by snowman2** [comment]
Minor tweak:

For example, if the semi-major axis length of the ellipsoid ~~is~~ defined by the grid mapping attribute semi_major_axis **disagrees with** the crs_wkt attribute (via the WKT SPHEROID[…​] element), the value of this attribute cannot be interpreted accurately.

**Comment by taylor13** [comment]
@snowman2 My poor brain can't seem to detect what exactly was tweaked.  Could you please point to the specific change made?

**Comment by taylor13** [comment]
Oh, I just saw that the crossed out "is".  Sorry.

**Comment by cameronsmith1** [comment]
I am OK with this (as amended).

**Comment by JonathanGregory** [comment]
Thanks for correcting my typo, @snowman2. I agree with Karl @taylor13's second point. It is a general statement. However, I think we're introducing unnecessary repetition. I appreciate that the modified text is in the pull request, but our guidelines are that we should discuss it as far as possible in the issue, so there's only one place to look to see the discussion. So I'm repeating the whole of paragraph and the previous one for context. I propose minor deletions for conciseness and to reduce repetition. How's this:

The crs_wkt attribute is intended to act as a supplement to other single-property CF grid mapping attributes (as described in Appendix F); it is not intended to replace those attributes. If data producers omit the single-property grid mapping attributes in favour of the compound crs_wkt attribute, software which cannot interpret crs_wkt will be unable to use the grid_mapping information. Therefore the CRS should be described as thoroughly as possible with the single-property attributes as well as by crs_wkt.

In cases where CRS property values can be represented by both a single-property grid mapping attribute and the crs_wkt attribute, the grid mapping should be provided, and if both are provided, the onus is on data producers to ensure that their property values are consistent. Therefore information from either one (or both) may be read in by the user without needing to check both. However, if the two values of a given property are different, the CRS information cannot be interpreted accurately and users should inform the provider so the issue can be addressed. For example, if the semi-major axis length of the ellipsoid is defined by the grid mapping attribute semi_major_axis disagrees with the crs_wkt attribute (via the WKT SPHEROID[…​] element), the value of this attribute cannot be interpreted accurately. Naturally if the two values are equal then no ambiguity arises.

Jonathan

**Comment by snowman2** [comment]
Sounds like a reasonable change to me.

Minor tweaks:

> compound crs_wkt attribute

I am thinking the `"compound"` word could be removed. 

>  the ellipsoid is defined by

I am thinking the `"is"` should be removed.

**Comment by JonathanGregory** [comment]
Thanks, @snowman2. I agree with both of those changes. Please could you update your pull request so it's the same text as above (with those two changes)?

If Karl @taylor13 and Philip @cameronsmith1 think that's OK still, we can count them as supporters, which means the proposal meets the conditions for acceptance. It will be accepted three weeks from now (3rd August) if there are no further concerns raised before then.

**Comment by snowman2** [comment]
Sounds good, just updated the PR (https://github.com/cf-convention/cf-conventions/pull/282/commits/9ad24ca7768416f71c8056fe14e294d8a7564a76).

**Comment by taylor13** [comment]
Yes, count me as a supporter.  Thanks.

**Comment by JonathanGregory** [comment]
Thanks, Karl

**Comment by cameronsmith1** [comment]
These changes look good to me.

**Comment by JonathanGregory** [comment]
There have been no further comments in the last three weeks and the required level of support has been reached so the proposal is accepted according to the rules. I have merged the pull request https://github.com/cf-convention/cf-conventions/pull/282/. Thanks, Alan @snowman2

**Comment by snowman2** [comment]
Thanks for assisting getting this proposal accepted @JonathanGregory 👍
